src/psyclone/domain/gocean/nodes/gocean_extract_node.py

```diff
@@ -141,66 +142,111 @@ def create_driver(self, input_list, output_list):
-                               .format(var_name))
+                               "{0}%ReadVariable(\"{1}\", {2})"
+                               .format(psy_data, var_name, local_name))
                 prog.add(call)
             elif is_input:
                 # Now must be input and output:
-                # First read the pre-variable
+                # First read the pre-variable (which is also allocated):
                 call = CallGen(prog,
-                               "psy_data%ReadVariable(\"{0}\", {0})"
-                               .format(var_name))
+                               "{0}%ReadVariable(\"{1}\", {2})"
+                               .format(psy_data, var_name, local_name))
                 prog.add(call)
                 # Then declare the post variable, and and read its values
                 # (ReadVariable will also allocate it)
-                decl = DeclGen(prog, "real", [var_name+post_suffix],
+                decl = DeclGen(prog, "real", [local_name+post_suffix],
                                dimension=":,:", kind="8", allocatable=True)
                 prog.add(decl)
                 call = CallGen(prog,
-                               "psy_data%ReadVariable(\"{0}\", {0})"
-                               .format(var_name+post_suffix))
+                               "{0}%ReadVariable(\"{1}{3}\", {2}{3})"
+                               .format(psy_data, var_name, local_name,
+                                       post_suffix))
                 prog.add(call)
             else:
                 # Now the variable is output only. We need to read the
-                # post variable in, and allocate the pre variable with
-                # the same size as the post
-                decl = DeclGen(prog, "real", [var_name+post_suffix],
+                # post variable in, and create and allocate a pre variable
+                # with the same size as the post
+                decl = DeclGen(prog, "real", [local_name+post_suffix],
                                dimension=":,:", kind="8", allocatable=True)
                 prog.add(decl)
                 call = CallGen(prog,
-                               "psy_data%ReadVariable(\"{0}\", {0})"
-                               .format(var_name+post_suffix))
+                               "{0}%ReadVariable(\"{1}{3}\", {2}{3})"
+                               .format(psy_data, var_name, local_name,
+                                       post_suffix))
                 prog.add(call)
-                decl = DeclGen(prog, "real", [var_name], kind="8",
+                decl = DeclGen(prog, "real", [local_name], kind="8",
                                dimension=":,:", allocatable=True)
                 prog.add(decl)
                 alloc = AllocateGen(prog, [var_name,
-                                           "mold={0}".format(var_name +
+                                           "mold={0}".format(local_name +
```

Comment:
I failed to spot this last time but really, AllocateGen should be modified to take a `mold` argument. I did think about whether this should be replaced by PSyIR but, in all likelihood, it's going to be a while before we support memory allocation. Please could you therefore create an issue and reference it here.

Reply:
I think we agreed for now to call it all grid_property :) But I added comments to explain this here.

Reply:
Hmm - is this good enough? That does not modify fparser, that seems to be consistent with the way fparser alloc statement is handled: you parse 'allocate(dummy)', then replace the items ('dummy' in this case) with the value you actually want.

```git diff f2pygen.py
diff --git a/src/psyclone/f2pygen.py b/src/psyclone/f2pygen.py
index 99fb01fc..43d829a3 100644
--- a/src/psyclone/f2pygen.py
+++ b/src/psyclone/f2pygen.py
@@ -793,12 +793,14 @@ def adduse(name, parent, only=False, funcnames=None):
 
 class AllocateGen(BaseGen):
     ''' Generates a Fortran allocate statement '''
-    def __init__(self, parent, content):
+    def __init__(self, parent, content, mold=None):
         '''
         :param parent: node to which to add this ALLOCATE as a child
         :type parent: :py:class:`psyclone.f2pygen.BaseGen`
         :param content: string or list of variables to allocate
         :type content: list of strings or a single string
+        :param mold: A string to be used as the 'mold' parameter of ALLOCATE.
+        :type mold: str or None.
 
         :raises RuntimeError: if `content` is not of correct type
         '''
@@ -814,6 +816,8 @@ class AllocateGen(BaseGen):
             raise RuntimeError(
                 "AllocateGen expected the content argument to be a str or"
                 " a list, but found {0}".format(type(content)))
+        if mold:
+            self._decl.items.append("mold={0}".format(mold))
         BaseGen.__init__(self, parent, self._decl)
```

Reply:
This is the corresponding change in the code creating the Allocate statement:
```
-                alloc = AllocateGen(prog, [var_name,
-                                           "mold={0}".format(local_name +
-                                                             post_suffix)])
+                alloc = AllocateGen(prog, [var_name],
+                                    mold="{0}".format(local_name +
+                                                      post_suffix))
```
I have not added this patch in yet, let me know if this is ok enough, or if you prefer a ticket with a more thorough solution (that will require changes to fparser as far as I can see)

Reply:
That's fine - that's all I was thinking of.

Reply:
OK, added, including a test-case.